mreg_cli/errorbuilder.py

```diff
@@ -0,0 +1,116 @@
+        return self.offset(self.CHAR_UNDERLINE * (end - start), start)
+
+    def get_suggestion(self, start: int) -> str:
+        """Get the suggestion part of the message."""
+        msg = self.SUGGESTION
+        if self.CHAR_SUGGESTION:
+            msg = f"{self.CHAR_SUGGESTION} {msg}"
+        return self.offset(msg, start)
+
+    def offset(self, s: str, n: int) -> str:
+        """Offset the string by n spaces."""
+        return " " * n + s
+
+    def build(self) -> str:
+        """Build the error message with an underline and suggestion."""
+        start, end = self.get_offset()
+        lines = [
+            self.command,
+            self.get_underline(start, end),
+            self.get_suggestion(start),
+        ]
+        return "\n".join(lines)
+
+    @abstractmethod
+    def get_offset(self) -> tuple[int, int]:
+        """Compute the start and end index of the underlined part of the command."""
+        pass
+
+
+class FilterErrorBuilder(ErrorBuilder):
+    """Error builder for commands with a filter."""
+
+    SUGGESTION = "Consider enclosing this part in quotes."
+
+    @staticmethod
+    def find_word_with_char_offset(command: str, char: str) -> tuple[int, int]:
+        """Find the start and end index of a word containing a specific character.
+
+        Cannot fail; will always return a tuple of two integers.
+        """
+        command = command.strip()
+        if not command:
+            return -1, -1
+
+        # Parse the command into parts
+        cmd = shlex.split(command, posix=False)  # keep quotes and backslashes
+        for part in cmd:
+            # Skip parts that contain quotes
+            # NOTE: This is a very naive check, and will not work for all cases
+            #      (e.g. escaped quotes)
+            if "'" in part or '"' in part:
+                continue
+
+            if char in part:
+                try:
+                    start, end = command.index(part), command.index(part) + len(part)
+                    if start > end or start >= len(command) or end >= len(command):
+                        raise ValueError(f"Invalid start and end values: {start}, {end}")
+                except ValueError as e:
+                    logger.error(
+                        "Failed to get index of char '%s' in part '%s' of command %s: %s",
+                        char,
+                        part,
+                        command,
+                        e,
+                    )
+                else:
+                    return start, end
+        return -1, -1
+
+    def get_offset(self) -> tuple[int, int]:  # noqa: D102 (missing docstring [inherit it from parent])
+        return self.find_word_with_char_offset(self.command, "|")
+
+
+def get_builder(command: str, exception: Exception | None = None) -> ErrorBuilder | None:
+    """Get the appropriate error builder for the given command."""
+    if "|" in command:
+        return FilterErrorBuilder(command, exception)
+    return None
```

Comment:
This should be refactored either to some sort of mapping lookup, or each Builder could define a `can_build()` (name TBD) class/static method that takes in the command and exception and returns `True`/`False`.